### PR TITLE
Document AOT Class Chains

### PR DESCRIPTION
**Pull request by dsouzai**

### CODE REVIEW

**Comment by dsouzai** [comment]
@mstoodle Could you please review :)

File: doc/compiler/AOTClassChains.md [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+to validate every single one. All this would effectively defeat the 
+purpose of AOT code, which is to facilitate fast application startup.
+
+Finding a particular `J9Class` in the current JVM instance requires 
+a Class Loader and a Class Name. The Class Name can be obtained from 
+the top romclass in the class chain. However, finding the Class Loader 
+is not as simple since they are just Java Objects created by a Java 
+application. OpenJ9 solves this problem by building has tables that 
+map a) Class Loaders to the first Class that loader loaded, and b) the 
+first class that loader loaded to the Class Loader Object. These tables
```
Comment (dsouzai) [inline doc/compiler/AOTClassChains.md]: @mstoodle To clarify, regarding the "first class that the loader loaded", what gets stored is the romclass (or its offset) right?

Comment (mstoodle) [reply doc/compiler/AOTClassChains.md]: what gets stored is an offset to the class chain for that class...

File: doc/compiler/AOTClassChains.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+has that particular class loaded would be extremely slow. If there were 
+multiple instances, then the AOT infrastructure would additionally need 
+to validate every single one. All this would effectively defeat the 
+purpose of AOT code, which is to facilitate fast application startup.
+
+Finding a particular `J9Class` in the current JVM instance requires 
+a Class Loader and a Class Name. The Class Name can be obtained from 
+the top romclass in the class chain. However, finding the Class Loader 
+is not as simple since they are just Java Objects created by a Java 
+application. OpenJ9 solves this problem by building has tables that
```
Comment (mstoodle) [inline doc/compiler/AOTClassChains.md]: typo: 'has' should be 'hash'

File: doc/compiler/AOTClassChains.md [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+purpose of AOT code, which is to facilitate fast application startup.
+
+Finding a particular `J9Class` in the current JVM instance requires 
+a Class Loader and a Class Name. The Class Name can be obtained from 
+the top romclass in the class chain. However, finding the Class Loader 
+is not as simple since they are just Java Objects created by a Java 
+application. OpenJ9 solves this problem by building has tables that 
+map a) Class Loaders to the first Class that loader loaded, and b) the 
+first class that loader loaded to the Class Loader Object. These tables
+are not persisted, it is built up via the Class Load Hook in the current
```
Comment (mstoodle) [inline doc/compiler/AOTClassChains.md]: 'it is' -> 'they are'

File: doc/compiler/AOTClassChains.md [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+Record, the JIT stores 1) the offset to the class chain for the `J9Class` 
+itself, as well as 2) the offset to the class chain for the first 
+loader that loaded the `J9Class`.
+
+An an AOT load, the AOT infrastructure takes 1) and adds to the 
+current address of the SCC in the current JVM instance to find the class 
+chain to be validated. This class chain provides the name of the class. 
+The AOT infrastructure then uses 2) to get the Class Loader Object in 
+the current JVM instance, which will likely give the correct Class Loader. 
+Using the Class Name, the Class Loader then provides a `J9Class` pointer
```
Comment (mstoodle) [inline doc/compiler/AOTClassChains.md]: maybe be more explicit that you ask the class loader for the J9Class pointer corresponding to the name (which may not yet be loaded...)

File: doc/compiler/AOTClassChains.md [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+current address of the SCC in the current JVM instance to find the class 
+chain to be validated. This class chain provides the name of the class. 
+The AOT infrastructure then uses 2) to get the Class Loader Object in 
+the current JVM instance, which will likely give the correct Class Loader. 
+Using the Class Name, the Class Loader then provides a `J9Class` pointer 
+which is used to perform the validation.
+
+## Subtleties
+For classes that are directly referenced in a method's code, for example 
+class `D` in `C.foo()` if `D.bar()` was a static call, there is a
```
Comment (mstoodle) [inline doc/compiler/AOTClassChains.md]: the wording to identify the class 'C' in a static call to D.bar() made by C.foo() here is a bit confusing, you may want to expand a bit to make it more clear

File: doc/compiler/AOTClassChains.md [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+already been loaded and is visible to class `C`, the the VM will return 
+the `J9Class` for class `D` directly, and this can be validated by a 
+constant pool index; the `validateClassFromConstantPool` function is used 
+to do this. This type of validation does **NOT** require validation via 
+the Class Chains.
+
+In other cases, there isn't any guarantee that a constant pool entry for 
+a particular class exists. In the example above, `d`, a reference of type 
+`D` may, maybe a reference to an object of type `D`, or it may be a 
+reference to some subclass of `D`, `E` extends `D`, that isnt' directly
```
Comment (mstoodle) [inline doc/compiler/AOTClassChains.md]: typo on the apostrophe

File: doc/compiler/AOTClassChains.md [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+
+In other cases, there isn't any guarantee that a constant pool entry for 
+a particular class exists. In the example above, `d`, a reference of type 
+`D` may, maybe a reference to an object of type `D`, or it may be a 
+reference to some subclass of `D`, `E` extends `D`, that isnt' directly 
+referenced by `C`. In this case, due to information from the IProfiler, 
+it might be known that `E` is the primary class of the `D` that's returned. 
+In that case, the JIT will insert a Profiled Inlined Guard that ensures 
+that `d.class == E` before executing the inlined code. In this scenario, 
+when performing an AOT load, the AOT infrastrcuture needs to not only
```
Comment (mstoodle) [inline doc/compiler/AOTClassChains.md]: typo: infrastructure

**Comment by dsouzai** [comment]
@mstoodle I made the requested changes.

File: doc/compiler/AOTClassChains.md [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+first class that loader loaded to the Class Loader Object. Regarding the
+"first class that loader loaded", what gets stored is an offset to the
+class chain for that class. These tables are not persisted, they are 
+built up via the Class Load Hook in the JIT in the current JVM instance. 
+When referencing a particular `J9Class` in a Relocation Record, the JIT 
+stores 1) the offset to the class chain for the `J9Class` 
+itself, as well as 2) the offset to the class chain for the first 
+loader that loaded the `J9Class`.
+
+An an AOT load, the AOT infrastructure takes 1) and adds to the
```
Comment (mstoodle) [inline doc/compiler/AOTClassChains.md]: found another typo: An an

File: doc/compiler/AOTClassChains.md [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+loader that loaded the `J9Class`.
+
+An an AOT load, the AOT infrastructure takes 1) and adds to the 
+current address of the SCC in the current JVM instance to find the class 
+chain to be validated. This class chain provides the name of the class. 
+The AOT infrastructure then uses 2) to get the Class Loader Object in 
+the current JVM instance, which will likely give the correct Class Loader. 
+It then asks the Class Loader to provde the `J9Class` pointer 
+corresponding to the Class Name. It is worth  noting that this `J9Class`
+ma not yet be loaded. The validatio nis then performed using this `J9Class`
```
Comment (mstoodle) [inline doc/compiler/AOTClassChains.md]: ma (may) and validatio n (validation)

**Comment by dsouzai** [comment]
@mstoodle Fixed the typos. Found a couple of others when I went through it again and fixed those as well.

**Comment by mstoodle** [comment]
Thanks @dsouzai !

**Comment by mstoodle** [comment]
Only changed documentation, so no further testing required before merging.

**Pull request merged**